### ISSUE TITLE
Can't copy Google documents to OneDrive

### ISSUE THREAD

**Issue by izzobot** [ncw/rclone#2832]
#### What is the problem you are having with rclone?

I have a daily sync that copies all of my google documents from Google Drive to OneDrive but for a while now it's not been working properly, no files are copied/synced, as you can see by the logs below.
It works fine if I transfer from Google Drive to local.
Please let me know if you need any more information or if I have misconfigured/misunderstood something.

Thank you.

#### What is your rclone version (output from `rclone version`)

rclone v1.45
- os/arch: linux/amd64
- go version: go1.11.2

(I have also tried the latest beta and I get the same result)

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)

4.18.20-unRAID GNU/Linux
Windows 10, 64 bit
(It occurs on both)

####  Which cloud storage system are you using? (eg Google Drive)

Google Drive and OneDrive

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)

`rclone sync "goo:Google Suite/Docs/document.docx" "one:temp/" -vv`

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)

```
rclone sync "goo:Google Suite/Docs/document.docx" "one:temp/" -vv
2018/12/11 00:43:50 DEBUG : rclone: Version "v1.45" starting with parameters ["rcloneorig" "--config" "/boot/config/plugins/rclone/.rclone.conf" "sync" "goo:Google Suite/Docs/document.docx" "one:temp/" "-vv"]
2018/12/11 00:43:50 DEBUG : Using config file from "/boot/config/plugins/rclone/.rclone.conf"
2018/12/11 00:43:55 INFO  : One drive root 'temp': Waiting for checks to finish
2018/12/11 00:43:55 INFO  : One drive root 'temp': Waiting for transfers to finish
2018/12/11 00:43:55 DEBUG : pacer: Rate limited, sleeping for 1.749245704s (1 consecutive low level retries)
2018/12/11 00:43:55 DEBUG : pacer: low level retry 1/10 (error googleapi: Error 403: Rate Limit Exceeded, rateLimitExceeded)
2018/12/11 00:43:55 DEBUG : pacer: Rate limited, sleeping for 2.148953723s (2 consecutive low level retries)
2018/12/11 00:43:55 DEBUG : pacer: low level retry 2/10 (error googleapi: Error 403: Rate Limit Exceeded, rateLimitExceeded)
2018/12/11 00:43:57 DEBUG : pacer: Resetting sleep to minimum 10ms on success
2018/12/11 00:43:58 INFO  : Waiting for deletions to finish
panic: src file size must be >= 0

goroutine 88 [running]:
github.com/ncw/rclone/backend/onedrive.(*Object).Update(0xc000114000, 0x139d9c0, 0xc00047a000, 0x1507d74d4630, 0xc00014a600, 0xc000544200, 0x1, 0x1, 0x0, 0x0)
        /home/travis/gopath/src/github.com/ncw/rclone/backend/onedrive/onedrive.go:1545 +0x260
github.com/ncw/rclone/backend/onedrive.(*Fs).Put(0xc0004945a0, 0x139d9c0, 0xc00047a000, 0x1507d74d4630, 0xc00014a600, 0xc000544200, 0x1, 0x1, 0xc000649e60, 0x8, ...)
        /home/travis/gopath/src/github.com/ncw/rclone/backend/onedrive/onedrive.go:715 +0x199
github.com/ncw/rclone/fs/operations.Copy(0x13b6900, 0xc0004945a0, 0x0, 0x0, 0xc00052a2b0, 0x10, 0x13b6080, 0xc00014a600, 0xc00055ef60, 0xc00049a500, ...)
        /home/travis/gopath/src/github.com/ncw/rclone/fs/operations/operations.go:295 +0x1299
github.com/ncw/rclone/fs/sync.(*syncCopyMove).pairCopyOrMove(0xc000030a80, 0xc0001de3c0, 0x13b6900, 0xc0004945a0, 0xc000030b58)
        /home/travis/gopath/src/github.com/ncw/rclone/fs/sync/sync.go:280 +0x23c
created by github.com/ncw/rclone/fs/sync.(*syncCopyMove).startTransfers
        /home/travis/gopath/src/github.com/ncw/rclone/fs/sync/sync.go:306 +0x9f
```

**Comment by Cnly** [comment]
@ncw I think I'm a little confused - in what cases does `(*Fs).Put` receive a source object with a negative size? Does it indicate streaming an object with an unknown size? In my memory OneDrive doesn't support that (need to check that again). Maybe we just need to change the `Panic` to an error returned?

**Comment by ncw** [comment]
If rclone doesn't know the size of an object it returns it as `-1`.  This is used in the google drive backend for google docs, and possibly elsewhere but it isn't very common.

So I think this needs 3 fixes

1. Make onedrive backend return an error if the size is < 0

2. In `operations.Copy`, make it call `operations.Rcat` if the input size is -1 which will buffer it on disk if necessary or use `PutStream` if available.

3. Require all backends to handle a file of size -1 sensibly in Put (either uploading it or giving an error) by adding a test to the integration tests.

Onedrive doesn't appear to support PutStream...

**Comment by Cnly** [comment]
@ncw I read a few backend implementations and it looks like some of them are aware of the `-1` size while others don't seem to have considered this case at all. And since we have the method `PutStream` which is designed to handle unknown-sized uploads, how about we make sure (and write in the docs) that `Put` and `Update` won't be called from outside a backend with a size < 0? If there's a file of unknown size, rclone could just call `PutStream` or `Rcat` instead.

**Comment by ncw** [comment]
@Cnly 
> I read a few backend implementations and it looks like some of them are aware of the `-1` size while others don't seem to have considered this case at all.

I think you'll find some of them will work as they don't need the length in advance.  Anyway that is why I suggested getting it in the integration tests so we can be consistent about it!

> And since we have the method `PutStream` which is designed to handle unknown-sized uploads, how about we make sure (and write in the docs) that `Put` and `Update` won't be called from outside a backend with a size < 0?

That is probably a good idea...

> If there's a file of unknown size, rclone could just call `PutStream` or `Rcat` instead.

That is what should be happening certainly!

**Comment by Cnly** [comment]
Sorry I've got a little bit busy recently. This bug has already have its critical part fixed, leaving some general changes to be made. For people who are affected by this bug, please try [this beta](https://beta.rclone.org/branch/v1.45-060-g2630a0a5-unknown-size-copy-2832-beta/) as a temporary fix which is the version compiled with the commit referenced above. It should prevent the panic and do the work correctly.

**Comment by Cnly** [comment]
This should have been fixed by #2947.
